In [ ]:
!curl https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet > data/yellow_tripdata_2023-01.parquet
!curl https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet > data/yellow_tripdata_2023-02.parquet

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 45.4M  100 45.4M    0     0  12.5M      0  0:00:03  0:00:03 --:--:-- 12.5M
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 32 45.5M   32 14.9M    0     0  8896k      0  0:00:05  0:00:01  0:00:04 8893k

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

# Question 1
Read the data for January. How many columns are there?

In [2]:
df_train = pd.read_parquet("data/yellow_tripdata_2023-01.parquet")
df_train

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee  
0                         2.5         0.00  
1                         2.5         0.00  
2                         2.5         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
3066761                   NaN          NaN  
3066762                   NaN          NaN  
3066763                   NaN          NaN  
3066764                   N

## Answer: 19 columns

# Question 2
Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

In [3]:
df_train.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
airport_fee                     float64
dtype: object

In [4]:
df_train["duration"] = df_train.tpep_dropoff_datetime - df_train.tpep_pickup_datetime
df_train["duration"] = df_train["duration"].apply(lambda tm: tm.total_seconds() / 60)
df_train

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee   duration  
0                         2.5         0.00   8.433333  
1                         2.5         0.00   6.316667  
2                         2.5         0.00  12.750000  
3                         0.0         1.25   9.616667  
4                         2.5         0.00  10.833333  
...                       ...          ...        ...  
3066761                   NaN          NaN  13.983333  
3066762                   NaN

In [5]:
df_train["duration"].std()

np.float64(42.59435124195458)

The standard deviation is: np.float64(42.59435124195458)
## Answer 42.59

# Question 3

Next, we need to check the distribution of the duration variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

In [6]:
total_records = df_train.shape[0]
total_records

3066766

In [7]:
df_train.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
airport_fee                     float64
duration                        float64
dtype: object

In [8]:
df_train = df_train[(df_train["duration"] >= 1) & (df_train["duration"] <= 60)]
df_train

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee   duration  
0                         2.5         0.00   8.433333  
1                         2.5         0.00   6.316667  
2                         2.5         0.00  12.750000  
3                         0.0         1.25   9.616667  
4                         2.5         0.00  10.833333  
...                       ...          ...        ...  
3066761                   NaN          NaN  13.983333  
3066762                   NaN

In [9]:
total_filtered = df_train.shape[0]
total_filtered

3009173

In [10]:
frac_no_outliers = (total_filtered * 100) / total_records
frac_no_outliers

98.1220282212598

## Answer: 98%


# Question 4 
Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

    - Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will label encode them)
    - Fit a dictionary vectorizer
    - Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

In [11]:
from sklearn.feature_extraction import DictVectorizer 

In [12]:
FEATURES_CAT = ['PULocationID','DOLocationID']

In [13]:
df_train[FEATURES_CAT] = df_train[FEATURES_CAT].astype(str)

/tmp/ipykernel_9057/3186627732.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_train[FEATURES_CAT] = df_train[FEATURES_CAT].astype(str)


In [14]:
df_train_vec = df_train[FEATURES_CAT].to_dict(orient='records')
df_train_vec

[{'PULocationID': '161', 'DOLocationID': '141'},
 {'PULocationID': '43', 'DOLocationID': '237'},
 {'PULocationID': '48', 'DOLocationID': '238'},
 {'PULocationID': '138', 'DOLocationID': '7'},
 {'PULocationID': '107', 'DOLocationID': '79'},
 {'PULocationID': '161', 'DOLocationID': '137'},
 {'PULocationID': '239', 'DOLocationID': '143'},
 {'PULocationID': '142', 'DOLocationID': '200'},
 {'PULocationID': '164', 'DOLocationID': '236'},
 {'PULocationID': '141', 'DOLocationID': '107'},
 {'PULocationID': '234', 'DOLocationID': '68'},
 {'PULocationID': '79', 'DOLocationID': '264'},
 {'PULocationID': '164', 'DOLocationID': '143'},
 {'PULocationID': '138', 'DOLocationID': '33'},
 {'PULocationID': '33', 'DOLocationID': '61'},
 {'PULocationID': '79', 'DOLocationID': '186'},
 {'PULocationID': '90', 'DOLocationID': '48'},
 {'PULocationID': '113', 'DOLocationID': '255'},
 {'PULocationID': '237', 'DOLocationID': '239'},
 {'PULocationID': '143', 'DOLocationID': '229'},
 {'PULocationID': '137', 'DOLocat

In [15]:
dv = DictVectorizer()
dv.fit_transform(df_train_vec)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 6018346 stored elements and shape (3009173, 515)>

## Answer: 515

# Question 5
Now let's use the feature matrix from the previous step to train a model.

    Train a plain linear regression model with default parameters, where duration is the response variable
    Calculate the RMSE of the model on the training data

What's the RMSE on train?

In [16]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error

In [17]:
X_train = dv.fit_transform(df_train_vec)
y_train = df_train["duration"].values

lr = LinearRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_train)
y_pred

array([11.52730355, 10.89778545, 11.32568417, ..., 11.73770134,
       12.70523274, 11.54226333], shape=(3009173,))

In [18]:
score = root_mean_squared_error(y_train,y_pred)
score

7.649261931416412

## Answer: 7.64

# Question 6
Now let's apply this model to the validation dataset (February 2023).

What's the RMSE on validation?

For the cleaning of the validation dataset, i will use the following function that compiles all the steps of the training set:

In [19]:
def read_dataframe(filename):
    df = pd.read_parquet(filename)
    df["duration"] = df_train.tpep_dropoff_datetime - df_train.tpep_pickup_datetime
    df.duration = df.duration.apply(lambda tm: tm.total_seconds() / 60)
    df = df[(df.duration >= 1) & (df.duration <= 60)]
    categorical = ['PULocationID', 'DOLocationID']
    df[categorical] = df[categorical].astype(str)
    return df

In [20]:
df_validate = read_dataframe("data/yellow_tripdata_2023-02.parquet")
df_validate

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
1               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
2               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
3               1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4               2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   
...           ...                  ...                   ...              ...   
2913950         2  2023-02-28 23:46:00   2023-03-01 00:05:00              NaN   
2913951         2  2023-02-28 23:26:02   2023-02-28 23:37:10              NaN   
2913952         2  2023-02-28 23:24:00   2023-02-28 23:38:00              NaN   
2913953         2  2023-02-28 23:03:00   2023-02-28 23:10:00              NaN   
2913954         2  2023-02-28 23:03:03   2023-02-28 23:12:51              NaN   

         trip_distance  RatecodeID store_and_fwd_flag PULocationID  \
0                 0.30         1.0                  N          142   
1                 0.00         1.0                  N           71   
2                 0.00         1.0                  N           71   
3                18.80         1.0                  N          132   
4                 3.22         1.0                  N          161   
...                ...         ...                ...          ...   
2913950           4.65         NaN               None          249   
2913951           2.47         NaN               None          186   
2913952           3.49         NaN               None          158   
2913953           2.13         NaN               None           79   
2913954           2.28         NaN               None          161   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                163             2         4.40   3.50      0.5        0.00   
1                 71             4        -3.00  -1.00     -0.5        0.00   
2                 71             4         3.00   1.00      0.5        0.00   
3                 26             1        70.90   2.25      0.5        0.00   
4                145             1        17.00   1.00      0.5        3.30   
...              ...           ...          ...    ...      ...         ...   
2913950          140             0        20.22   0.00      0.5        4.84   
2913951           79             0        13.66   0.00      0.5        2.65   
2913952          143             0        17.64   0.00      0.5        0.00   
2913953          162             0        13.56   0.00      0.5        2.63   
2913954          140             0        14.89   0.00      0.5        3.78   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0          9.40   
1                 0.0                   -1.0         -5.50   
2                 0.0                    1.0          5.50   
3                 0.0                    1.0         74.65   
4                 0.0                    1.0         25.30   
...               ...                    ...           ...   
2913950           0.0                    1.0         29.06   
2913951           0.0                    1.0         20.31   
2913952           0.0                    1.0         21.64   
2913953           0.0                    1.0         20.19   
2913954           0.0                    1.0         22.67   

         congestion_surcharge  Airport_fee   duration  
0                         2.5         0.00   8.433333  
1                         0.0         0.00   6.316667  
2                         0.0         0.00  12.750000  
3                         0.0         1.25   9.616667  
4                         2.5         0.00  10.833333  
...                       ...          ...        ...  
2913950                   NaN          NaN   2.883333  
2913951                   NaN          NaN   7.700000

In [21]:
y_val = df_validate["duration"].values

In [22]:
df_val_vec = df_validate[FEATURES_CAT].to_dict(orient='records')
len(df_val_vec)

2858634

In [23]:
X_val = dv.transform(df_val_vec)

In [24]:
y_val_pred = lr.predict(X_val)

In [25]:
score_val = root_mean_squared_error(y_val,y_val_pred)
score_val

11.87337423387558

## Answer 11.81